Move engines to Warp

gakonst opened the pull request:
As title, same principle as https://github.com/interledger-rs/interledger-rs/pull/420

gakonst committed:
feat(engine): use warp instead of tower_web

gakonst committed:
chore: cargo fmt + clippy

emschwartz commented on src/api.rs:
```diff
@@ -1,6 +1,11 @@
+/// # Settlement Engine API
+///
+/// Tower_Web service which exposes settlement related endpoints as described in RFC536,
```
This comment should be updated

emschwartz commented on src/api.rs:
```diff
@@ -9,145 +14,235 @@ use hyper::{Response, StatusCode};
+// Helper function that returns any idempotent data that corresponds to a
+// provided idempotency key. It fails if the hash of the input that
+// generated the idempotent data does not match the hash of the provided input.
+fn check_idempotency<S>(
```
If we're going to make this crate depend on `interledger-api`, what do you think about having that crate export shared functionality like this (also behind a feature flag)?

gakonst replied:
I agree with that.

emschwartz reviewed (approved):
LGTM. @dora-gt should review as well

gakonst committed:
fix(api): recover properly from idempotency errors

also use idempotent and leftovers store implementations from interledger_settlement

gakonst committed:
fix(engine): make engines return proper errors

gakonst committed:
chore: cargo fmt

emschwartz commented on Cargo.toml:
```diff
@@ -7,20 +7,20 @@ license = "Apache-2.0"
 futures = "0.1.25"
 interledger-service = "^0.2.2-alpha.1"
-interledger-settlement = { git = "https://github.com/interledger-rs/interledger-rs", branch = "master" }
+interledger-settlement = { git = "https://github.com/interledger-rs/interledger-rs", branch = "gakonst/settlement-warp" }
```
This does make me think that `interledger-settlement` should be included in this repo

gakonst replied:
Agreed.

emschwartz reviewed (commented):
Gah I hate when I review something and then it doesn't get submitted and stays pending. Sorry for the delay

dora-gt commented on src/api.rs:
```diff
@@ -1,194 +1,174 @@
+    let account_id = warp::path("accounts").and(warp::path::param2::<String>()); // account_id
+
+    // POST /accounts/ (optional idempotency-key header)
+    // Body is a Vec<u8> object
```
Is this comment right? It seems `.and(warp::body::json())`.

dora-gt commented on src/api.rs:
```diff
@@ -284,68 +262,37 @@ mod tests {
+        let api = create_settlement_engine_filter(engine, store.clone());
+
+        let settlement_call = move |id, amount, scale| {
+            warp::test::request()
```
👀

I'll consider doing this next time with `TestStore`.

dora-gt reviewed (approved):
LGTM, just left a comment.
